충남대 BE_이은경 6주차 과제 (3단계)

pkyung opened the pull request:
## 과제 진행 사항
### 포인트
- [x] member 필드에 point 추가, 필요한 메서드 추가
- [x] point api cr 생성
- [x] 관리자 페이지에서 포인트 생성 및 조회 가능

주문 과정에서 usedPoint 필드를 추가하여 주문 시에 멤버가 갖고 있는 포인트를 차감합니다. 
포인트 충전은 관리자 페이지에서 진행합니다.

pkyung committed:
Chore: 5주차 코드 이동

pkyung committed:
Chore: docs 수정

pkyung committed:
Docs: 기능 목록 작성 - step1

pkyung committed:
Feat: api 통일

1. authController에 /api 추가함
2. OptionResponse productId 제거
3. Option에 toDto productId 제거
4. wishes 조회 시 product 반환이 아닌 wishResponse 반환하도록 수정
5. wishResponse에는 wishId를 추가함
6. wish 삭제 시 wishId에 대해서 삭제되도록 수정함 (기존에 productId에
   대해서 수정했음)

pkyung committed:
Feat: 수정된 api에 대한 관리자 페이지 수정

1. auth.js에는 /api 추가함
2. script.js에는 category로 반환하는 productResponse에 대한 수정 작업합
3. wish.js에는 delete route 수정과 wish 반환값 수정함

pkyung committed:
Feat: 상품 카테고리 별 조회 추가

기존의 상품을 조회하는 api는 /all 을 추가하고 카테고리 별로 조회하는
api는 쿼리 파라미터로 categoryId를 받아서 처리했다.

pkyung committed:
Docs: 기능 목록 구현 - step1

pkyung committed:
Fix: order 과정에서 카카오 로그인이 아닌 경우 처리

기존의 코드에는 카카오 로그인이 아닌 경우, 주문 메시지 처리 과정에서
에러가 났다.
member 객체에 accessToken이 null인지 확인 후, null 이면 메시지 전송
과정은 하지 않도록 처리했다.

pkyung committed:
Merge branch 'pkyung' into main

pkyung committed:
Fix: route 수정

pkyung committed:
Fix: page 직렬화 반환 수정

pkyung committed:
Test: 테스트 api 수정에 따라 수정

pkyung committed:
Fix: cors 수정 및 data 추가 sql

pkyung committed:
Fix: sql 테스트 통과가 안되어 파일 삭제

pkyung committed:
Fix: categories url 실수 수정

pkyung committed:
Refactor: 디미터 법칙 위반된 것을 수정함

Wish 도메인에 isOwnedBy 메서드를 추가하여 getId와 같은 메서드를 직접
호출하지 않고 비교했습니다.

pkyung committed:
Merge branch 'pkyung' into main

pkyung committed:
Refactor: 오류 해결

pkyung committed:
Test: CORS 테스트 추가

pkyung committed:
Docs: 기능 목록 작성 - step3

pkyung committed:
Docs: 기능 목록 수정

pkyung committed:
Feat: member 필드에 point 추가, 필요한 메서드 추가

1. member 필드에 int형의 point 필드를 추가했다.
2. 포인트를 충전하는 chargePoint 메서드를 구현했다.
3. 포인트를 차감하는 subtractPoint를 구현했다. 이 때 차감한 포인트가
   0보다 작다면 IllegalArgumentException을 반환하며 이때 400 bad request
   에러를 반환한다.

pkyung committed:
Feat: point api cr 생성

1. pointRequest와 pointResponse를 record로 생성했다.
2. pointController에 포인트의 충전과 조회 메서드를 추가했다.
3. 주문 시에 usedPoint 필드를 추가하여 주문할 때 사용한 포인트를
   차감하도록 구현했다.

pkyung committed:
Feat: 관리자 페이지에서 포인츠 충전 및 차감

메인 페이지에서 로그인 후에 포인트를 충전할 수 있게 구현했다.

pkyung committed:
Docs: 기능 목록 구현 - step3

pkyung committed:
Merge branch 'pkyung' into main

ksy90101 commented on src/main/java/gift/controller/PointController.java:
```diff
@@ -0,0 +1,32 @@
+    @PostMapping
+    public ResponseEntity<Void> chargePoint(@RequestBody PointRequest request, @LoginMember Member member) {
+        pointService.pointCharge(request, member);
+        return ResponseEntity.status(HttpStatus.CREATED).build();
```
포인트를 변경하는것 같은데 POST가 맞을가요?

pkyung replied:
put으로 변경하겠습니다.

ksy90101 replied:
아울러 과연 points가 REST API의 관점에서 리소스인가를 고민해보면 좋을거 같습니다.

member의 상태이니 member가 리소스이고, 그에 대한 상태가 될거 같은데.

members/point와 같은 URL은 고민해보셨을까요?

pkyung replied:
저 또한 포인트가 멤버의 필드로 들어가기에 url이 멘토님께서 말씀하신 것처럼 /api/members/points가 되어야하지 않을까 라는 생각을 했었습니다. PointService에서도 memberRepository만 가져오기에 memberService에서 담당할 수 있지 않나? 라는 의문도 들었습니다. 
팀원들과 api 명세를 하는 과정에서 이렇게 정해졌던 것 같습니다. 잘 못 생각했던 것 같습니다.

ksy90101 replied:
정답이 존재하진 않으니깐요~
팀원들과 소통이 되었다면 괜찮을거 같습니다~

ksy90101 commented on src/main/java/gift/dto/OrderRequest.java:
```diff
@@ -11,6 +11,7 @@ public record OrderRequest(
         @NotBlank
-        String message
+        String message,
+        int usedPoint
```
여기도 validate가 필요할거 같습니다.

pkyung replied:
네 추가하겠습니다.

ksy90101 commented on src/main/java/gift/dto/PointRequest.java:
```diff
@@ -0,0 +1,6 @@
+package gift.dto;
+
+public record PointRequest(
+        int point
```
여기도 validate가 필요할거 같습니다.

ksy90101 commented on src/main/java/gift/service/PointService.java:
```diff
@@ -0,0 +1,31 @@
+        this.memberRepository = memberRepository;
+    }
+
+    public int pointCharge(PointRequest request, Member member) {
```
여기서는 DTO로 하지 않은 이유가 무엇인가요?

pkyung replied:
controller에서 반환을 하지 않아서 그렇게 했는데 생각해보니 void로 해도 될 것 같습니다!

ksy90101 reviewed (changes requested):
마지막이네요 👍 
세세한 부분에 대한 리뷰를 남겼으니 한번 확인해주세요~

pkyung committed:
Refactor: 멘토님 피드백 수정

1. dto validate 추가했다.
2. service 반환값 void로 설정했다.
3. 포인트를 변경하는 controller에서 postMapping을 putmapping으로
   수정했다.

ksy90101 commented on src/main/java/gift/model/Member.java:
```diff
@@ -56,4 +61,20 @@ public boolean matchPassword(String password) {
+        this.point = remainingPoint;
+    }
+
+    public void chargePoint(int usingPoint) {
```
여기서 usingPoint가 음수가 들어오면 어떻게 될까요?

pkyung replied:
이 메서드를 사용하는 부분에서 validation을 설정하긴했으나 여기서도 처리하도록 수정하겠습니다.

ksy90101 commented on src/main/java/gift/model/Member.java:
```diff
@@ -12,25 +13,29 @@ public class Member {
     @Column(nullable = false)
     private String password;
     private String accessToken;
+    private int point;
```
마지막으로 한번 Point를 VO로 만들어 보면 어떨까요?

pkyung replied:
네 만들어보겠습니다.

pkyung replied:
네 만들어보겠습니다.

ksy90101 commented on src/main/resources/static/js/point.js:
```diff
@@ -0,0 +1,46 @@
+    fetch('/api/points', {
+        method: 'GET',
+        headers: {
+            'Authorization': `Bearer ${token}`
```
여기도 Content-Type이 있어야 하지 않을까요?

pkyung replied:
네 추가하겠습니다.

ksy90101 commented on src/main/resources/static/js/point.js:
```diff
@@ -0,0 +1,46 @@
+    console.log(point)
+    const token = getAuthToken();
+
+    fetch('/api/points', {
+        method: 'POST',
+        headers: {
+            'Authorization': `Bearer ${token}`,
+            'Content-Type': 'application/json'
+        },
+        body: JSON.stringify({point})
+    }).then(response => {
```
한번 fetch하는걸 공통 로직으로 만들어보면 어떨까요?
중복 코드들이 존재하는거 같아요.

pkyung replied:
도전해봤는데 쉽지 않았습니다...

ksy90101 commented on src/main/java/gift/dto/PointRequest.java:
```diff
@@ -0,0 +1,9 @@
+
+import jakarta.validation.constraints.Min;
+
+public record PointRequest(
```
디폴트 값을 지정해봐도 좋을거 같긴 하네요.

pkyung replied:
네 지정해보겠습니다.

ksy90101 commented on src/main/java/gift/controller/PointController.java:
```diff
@@ -0,0 +1,33 @@
+
+    @PutMapping
+    public ResponseEntity<Void> chargePoint(@Valid @RequestBody PointRequest request, @LoginMember Member member) {
+        pointService.pointCharge(request, member);
+        return ResponseEntity.status(HttpStatus.OK).build();
```
내려줄 데이터가 없을땐 204 no_content를 많이 사용합니다.

pkyung replied:
네 수정하겠습니다!

pkyung replied:
넵 확인했습니다.

ksy90101 reviewed (changes requested):
마지막 피드백이 될거 같네요~
간단한 피드백을 남겼으니 확인해주세요!

pkyung committed:
Refactor: 멘토님 피드백 수정

1. point 충천 시 음수값 일 때 exception 반환했다.
2. point를 VO로 생성했다.
3. content-type 추가했다.
4. pointRequest 디폴트 값 지정했다.
5. put 응답값 204 반환했다.

ksy90101 commented on src/main/java/gift/model/Point.java:
```diff
@@ -0,0 +1,45 @@
+@Embeddable
+@Access(AccessType.FIELD)
+public class Point {
+    @Column(name = "point_value")
+    private int value;
```
왜 point_value가 되었는지... ㅎㅎㅎ

ksy90101 replied:
VO는 불변을 유지하는편이라서 final을 쓰는게 좋을거 같아요.

ksy90101 commented on src/main/java/gift/model/Point.java:
```diff
@@ -0,0 +1,45 @@
+            throw new IllegalArgumentException("사용된 포인트가 음수일 수 없습니다.");
+        }
+
+        int remainingPoint = this.value - usedPoint;
+        if (remainingPoint < 0) {
+            throw new IllegalArgumentException("사용된 포인트가 갖고 있는 포인트보다 큰 값일 수 없습니다.");
+        }
+
+        this.value = remainingPoint;
+    }
+
+    public void chargePoint(int usingPoint) {
+        if (usingPoint < 0) {
+            throw new IllegalArgumentException("사용할 포인트가 음수일 수 없습니다.");
+        }
+        this.value += usingPoint;
```
새로운 VO인 new Point를 쓰는게 더 좋을거 같습니다.

ksy90101 reviewed (approved):
차후 VO에 대해서 좀더 고민해보시면 좋을거 같습니다~
그동안 고생 많으셨어요!
